### PR TITLE
RPATH support using ld wrapper (WIP)

### PR DESCRIPTION
**Pull request by rjeschmi**
I am working to getting this working to spec, but I thought I'd start the PR with a version using the wrapper. It works well enough, but I haven't removed the module deps yet (so LD_LIBRARY_PATH is still populated)

### CODE REVIEW

File: easybuild/tools/options.py [hunk 1]
```diff
@@ -433,6 +433,16 @@ def package_options(self):
         self.log.debug("package_options: descr %s opts %s" % (descr, opts))
         self.add_group_parser(opts, descr)
 
+    def rpath_options(self):
+        # rpath Options
+        descr = ("RPATH options", "changes linking to include rpaths")
+
+        opts = OrderedDict({
+            'rpath': ("Enable RPATH support", None, 'store_true', False),
```
Comment (boegel) [inline easybuild/tools/options.py]: do you expect other options here?

if not, this shouldn't be an option group by itself?

File: easybuild/framework/easyblock.py [hunk 2]
```diff
@@ -1523,6 +1524,12 @@ def prepare_step(self):
         # prepare toolchain: load toolchain module and dependencies, set up build environment
         self.toolchain.prepare(self.cfg['onlytcmod'], silent=self.silent)
 
+        self.log.debug("prepare_step: PATH %s" % os.environ['PATH'])
+        if build_option('rpath'):
+            # Setup the environment and copy wrapper script into path
+            prepare_ld_wrapper()
```
Comment (boegel) [inline easybuild/framework/easyblock.py]: hmm, shouldn't we do this in the `toolchain.prepare()` itself? keep things nicely together?

File: easybuild/tools/rpath.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+"""A module to setup and use an ld wrapper to support RPATH"""
```
Comment (boegel) [inline easybuild/tools/rpath.py]: missing license header + author

File: easybuild/tools/rpath.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+import os
+import copy
+import shutil
+import stat
+import tempfile
+from vsc.utils import fancylogger
+from easybuild.tools.filetools import adjust_permissions
+
+_log = fancylogger.getLogger('tools.package')
```
Comment (boegel) [inline easybuild/tools/rpath.py]: not package, rpath

File: easybuild/tools/rpath.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+orig_os_environ = copy.deepcopy(os.environ)
```
Comment (boegel) [inline easybuild/tools/rpath.py]: we have copies of the original environment in various places already, no need to keep another?

File: easybuild/tools/rpath.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+# RPATH is linux only
+ld_wrapper_script_loc = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "scripts", "ld_wrapper.sh"))
```
Comment (boegel) [inline easybuild/tools/rpath.py]: this is a constant, so name it `LD_WRAPPER_SCRIPT_LOC`

File: easybuild/tools/rpath.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    os.environ['EB_LD_FLAG'] = "1"
+    if DEBUG:
+        os.environ['EB_LD_VERBOSE'] = 'true'
+
+
+def teardown_ld_wrapper():
+    """
+    Return environment
+    """
+    os.environ = copy.deepcopy(orig_os_environ)
```
Comment (boegel) [inline easybuild/tools/rpath.py]: the environment is reset for every new build already, I don't think we need to worry about this?

File: easybuild/tools/rpath.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def check_rpath_support():
```
Comment (boegel) [inline easybuild/tools/rpath.py]: not called anywhere?

File: easybuild/tools/rpath.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    """
+    _log.debug("rpath: preparing ld wrapper script")
+    wrapper_dir = tempfile.mkdtemp(prefix='eb-ldwrapper-')
+    # copy wrapper script from framework
+    wrapper_ld = os.path.join(wrapper_dir, "ld")
+    wrapper_ld_gold = os.path.join(wrapper_dir, "ld.gold")
+    shutil.copy(ld_wrapper_script_loc, wrapper_ld)
+    shutil.copy(ld_wrapper_script_loc, wrapper_ld_gold)
+    adjust_permissions(wrapper_ld, stat.S_IXUSR, add=True)
+    adjust_permissions(wrapper_ld_gold, stat.S_IXUSR, add=True)
```
Comment (boegel) [inline easybuild/tools/rpath.py]:
```python
for linker in ['ld', 'ld.gold']:
    wrapper = os.path.join(wrapper_dir, linker)
    shutil.copy(ld_wrapper_script_loc, wrapper)
    adjust_permissions(wrapper, stat.S_IXUSR, add=True)
```

**Comment by boegel** [comment]
@rjeschmi: this looks pretty basic (good!), but I think wrapper `ld` is not what we plan to do in the end, right?

**Comment by rjeschmi** [comment]
I'll probably leave it as an "option"

I feel like having a wrapper available with some more generic hooks is a good thing and might lead to helpful debugging or special case handling in the future.

The way it is injected makes it low risk for inclusion even if it is not the default method

Ultimately for some cases this might work better than setting environment variables and is a good option to have around.

File: easybuild/scripts/ld_wrapper.sh [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+## More aggressive rpath-ing
+    exclude_lib_paths=("/tmp" "/opt")
+  fi
+
+  L=""
+  lib_array=()
+  dir=`pwd`
+  for (( i=${#@}; i >= 0; i-- )); do
+    if [[ ${!i} == "--enable-new-dtags" ]]; then
+	## we are removing this flag if passed as it creates a copy of rpath to runpath.
```
Comment (ocaisa) [inline easybuild/scripts/ld_wrapper.sh]: Aren't we at the mercy of what the default is set to here? We probably should (always) be adding `--disable-new-dtags` to the arguments as well as removing `--enable-new-dtags`

File: easybuild/framework/easyblock.py (hunk 2 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: see `easybuild/tools/toolchain/toolchain.py:prepare`

**Comment by boegel** [comment]
TODO:

* add sanity check (using pyreadelf)
* allow rpath methods: env, ld wrapper, compilers wrappers (multiple options at the same)
* no `$LD_LIBRARY_PATH` in modules (and check for it when loading dep modules)

**Pull request merged**